include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,425 @@
+// Copyright 2020 Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+
+#include <boost/math/special_functions/prime.hpp>
+#include <boost/assert.hpp>
+#include <vector>
+#include <iterator>
+#include <cmath>
+#include <thread>
+#include <memory>
+#include <future>
+#include <numeric>
+#include <iostream>
+#include <algorithm>
+#include <execution>
+
+namespace boost::math { namespace detail
+{
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Integer, class Container>
+void linear_sieve(Integer upper_bound, Container &resultant_primes)
+{
+    size_t least_divisors_size{static_cast<size_t>(upper_bound + 1)};
+    std::unique_ptr<Integer[]> least_divisors{new Integer[least_divisors_size]{0}};
+
+    for (Integer i{2}; i <= upper_bound; ++i)
+    {
+        if (least_divisors[i] == 0)
+        {
+            least_divisors[i] = i;
+            resultant_primes.emplace_back(i);
+        }
+
+        for (size_t j{}; j < least_divisors_size; ++j)
+        {
+            if (j >= resultant_primes.size())
+            {
+                break;
+            }
+
+            else if (resultant_primes[j] > least_divisors[i])
+            {
+                break;
+            }
+
+            else if (i * resultant_primes[j] > upper_bound)
+            {
+                break;
+            }
+
+            else
+            {
+                least_divisors[i * resultant_primes[j]] = resultant_primes[j];
+            }
+        }
```

Comment:
Just a code simplification here, but those three conditions that result in a `break`, they could all go in the second clause of the `for` loop? I mean, `j < least_divisors && j < resultant_primes.size() && ... etc`.
Another way to simplify this with more abstraction would be to make a zip iterator over the two data structures.